fcm/fcm.go

```diff
@@ -0,0 +1,199 @@
+	MessageID   string `json:"message_id"`
+	TimeToLive  int    `json:"time_to_live"`
+}
+
+func NewFCM(clientId string, proxyName string, fcmServerBindUrl string, FcmSubscribe func(string) (interface{}, error), notifications chan<- notification.PrinterNotification) (*FCM, error) {
+	f := FCM{
+		fcmServerBindUrl,
+		"",
+		time.Time{},
+		clientId,
+		proxyName,
+		0,
+		FcmSubscribe,
+		notifications,
+		make(chan struct{}),
+		make(chan struct{}),
+	}
+	return &f, nil
+}
+
+//  get token from GCP and connect to FCM.
+func (f *FCM) Init()  {
+	iidToken := f.GetToken()
+	if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+		for err != nil {
+			log.Errorf("FCM restart failed, will try again in 10s: %s", err)
+			time.Sleep(10 * time.Second)
+			err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+		}
+		log.Info("FCM conversation restarted successfully")
+	}
+
+	go f.KeepFcmAlive()
+}
+
+// Quit terminates the FCM conversation so that new jobs stop arriving.
+func (f *FCM) Quit() {
+	// Signal to KeepFCMAlive.
+	close(f.quit)
+}
+
+// Fcm notification listener
+func (f *FCM) ConnectToFcm(fcmNotifications chan<- notification.PrinterNotification, iidToken string, dead chan<- struct{}, quit chan<- struct{}) (error){
+	log.Debugf("Connecting to %s?token=%s", f.fcmServerBindURL, iidToken)
+	resp, err := http.Get(fmt.Sprintf("%s?token=%s", f.fcmServerBindURL, iidToken))
+	if err != nil {
+		// failed for ever no need to retry
+		quit <- struct {}{}
+		return err
+	}
+	if resp.StatusCode == 200 {
+		reader := bufio.NewReader(resp.Body)
+		go func (){
+		for {
+			line, err := reader.ReadBytes('\n')
+			if err == nil || err == io.EOF {
+				printerId := GetPrinterID(string(line))
+				if printerId != "" {
+					pn := notification.PrinterNotification{printerId, notification.PrinterNewJobs}
+					go func() {
+						fcmNotifications <- pn
+					}()
+				}
+				if err == io.EOF {
+					log.Info("DRAIN message received, client reconnecting.")
+					dead <- struct{}{}
+					break
+				}
+			} else {
+				// stop listening unknown error happened.
+				log.Errorf("%v", err)
+				quit <- struct{}{}
+				break
+			}
+		}
+	}()
+	}
+	return nil
+}
+
+// Restart FCM connection when lost.
+func (f *FCM) KeepFcmAlive() {
+	for {
+		select {
+		case <-f.dead:
+			iidToken := f.GetToken()
+			log.Error("FCM conversation died; restarting")
+			if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+				for err != nil {
+					log.Errorf("FCM connection restart failed, will try again in 10s: %s", err)
+					time.Sleep(10 * time.Second)
+					err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+				}
+				log.Info("FCM conversation restarted successfully")
+			}
+
+		case <-f.quit:
+				log.Info("Fcm client Quitting ...")
+				// quitting keeping alive
+				return
+		}
+	}
+}
+// Returns cached token and Refresh token if needed.
+func (f *FCM) GetToken() (string){
```

Comment:
nit: a space between ')' and '{' for all functions below.

maybe run a lint to fix for all related?

Reply:
Done.